ProcessLib/ThermoMechanics/ThermoMechanicsFEM.h

```diff
@@ -0,0 +1,545 @@
+#include "ProcessLib/Deformation/LinearBMatrix.h"
+#include "ProcessLib/LocalAssemblerInterface.h"
+#include "ProcessLib/LocalAssemblerTraits.h"
+#include "ProcessLib/Parameter/Parameter.h"
+#include "ProcessLib/Utils/InitShapeMatrices.h"
+
+#include "ThermoMechanicsProcessData.h"
+
+namespace ProcessLib
+{
+namespace ThermoMechanics
+{
+template <typename BMatricesType, typename ShapeMatrixType, int DisplacementDim>
+struct IntegrationPointData final
+{
+    explicit IntegrationPointData(
+        MaterialLib::Solids::MechanicsBase<DisplacementDim>& solid_material)
+        : solid_material(solid_material),
+          material_state_variables(
+              solid_material.createMaterialStateVariables())
+    {
+    }
+
+#if defined(_MSC_VER) && _MSC_VER < 1900
+    // The default generated move-ctor is correctly generated for other
+    // compilers.
+    explicit IntegrationPointData(IntegrationPointData&& other)
+        : b_matrices(std::move(other._b_matrices)),
+          sigma(std::move(other._sigma)),
+          sigma_prev(std::move(other._sigma_prev)),
+          eps(std::move(other._eps)),
+          eps_m(std::move(other._eps_m)),
+          eps_m_prev(std::move(other._eps_m_prev)),
+          solid_material(other._solid_material),
+          material_state_variables(std::move(other._material_state_variables)),
+          C(std::move(other._C)),
+          integration_weight(std::move(other.integration_weight)),
+    {
+    }
+#endif  // _MSC_VER
+
+    typename ShapeMatrixType::NodalRowVectorType N;
+    typename ShapeMatrixType::GlobalDimNodalMatrixType dNdx;
+    typename BMatricesType::BMatrixType b_matrices;
+    typename BMatricesType::KelvinVectorType sigma, sigma_prev;
+    typename BMatricesType::KelvinVectorType eps;
+    typename BMatricesType::KelvinVectorType eps_m, eps_m_prev;
+
+    MaterialLib::Solids::MechanicsBase<DisplacementDim>& solid_material;
+    std::unique_ptr<typename MaterialLib::Solids::MechanicsBase<
+        DisplacementDim>::MaterialStateVariables> material_state_variables;
+
+    typename BMatricesType::KelvinMatrixType C;
+    double integration_weight;
+
+    void pushBackState()
+    {
+        eps_m_prev = eps_m;
+        sigma_prev = sigma;
+        material_state_variables->pushBackState();
+    }
+
+    static const int kelvin_vector_size =
+        KelvinVectorDimensions<DisplacementDim>::value;
+    using Invariants = MaterialLib::SolidModels::Invariants<kelvin_vector_size>;
+
+    template <typename DisplacementVectorType>
+    void updateConstitutiveRelation(double const t,
+                                    SpatialPosition const& x_position,
+                                    double const dt,
+                                    DisplacementVectorType const& u,
+                                    double const thermal_strain)
+    {
+        eps.noalias() = b_matrices * u;
+        eps_m.noalias() = eps - thermal_strain * Invariants::identity2;
```

Comment:
Make a comment that this assumes isotropic thermal expansion tensor ... this will need to be generalized to anisotropy later (not now)